docs/edsl.md

```diff
@@ -0,0 +1,739 @@
+
+```c++
+Tensor max_pool_1d(const Tensor& I) {
+  TensorDim N;
+  TensorIndex i, j;
+  I.bind_dims(N);
+  auto O = TensorOutput(N / 2);
+  O(i) >= I(2 * i + j);
+  O.add_constraints({j<2});
+  return O;
+}
+```
+
+Something important to note here is that while we wrote `j < 2`, this constraint
+actually means `0<= j < 2`. Constraints are always bounded below by `0`.
+(Without a constraint, however, index variables may still be negative: the
+original code included e.g. `i = 1`, `j = -1` as valid index pair.)
+
+We determined the Tile code for this example by starting from imperative code,
+but this Tile code is still very similar to mathematical notation, and we could
+have started there instead:
+
+```math
+\Large
+\textcolor{red}{O[n]}
+\textcolor{yellow}{=}
+\textcolor{green}{\max}\textcolor{magenta}{_{0 \leq j < 2}}
+\textcolor{cyan}{I[2i + j]}
+```
+
+```math
+\Large
+\texttt{
+  if (\textcolor{magenta}{j < 2}) \{{
+    \textcolor{red}{O(n)}
+    \textcolor{green}{>=}
+    \textcolor{cyan}{I(2 * i + j)};
+  \}}
+}
+```
+
+This Tile code handles odd values of `N` by rounding down the output tensor
+size. You may instead want to round up the output tensor size and use a smaller
+pool at the edge. This can be accomplished by simply adjusting the size of `O`:
+
+```c++
+Tensor max_pool_1d(const Tensor& I) {
+  TensorDim N;
+  TensorIndex i, j;
+  I.bind_dims(N);
+  auto O = TensorOutput((N + 1) / 2);
+  if (j < 2) {
+    O(i) >= I(2 * i + j);
+  }
+  return O;
+}
+```
+
+No special handling is needed for the case `i = (N - 1) / 2`, `j = 1`; this is
+out of range for `I` and so is ignored by Tile, which is exactly the intended
+behavior.
+
+### Valid Indices
+
+When discussing contractions, we've mentioned that they accumulate over "all
+valid indices". Hopefully the significance of this has been clear for the
+specific examples we've looked at, but to write complex or novel code it helps
+to have a precise understanding of what is meant by "valid indices".
+
+First, index validity is determined for a full set of index variables: `j = 1`
+is not valid or invalid as a standalone index value, but may be part of a valid
+or invalid set of index variables. For example, in the code:
+
+```c++
+I.bind_dims(N);
+auto O = TensorOutput((N + 1) / 2);
+O(i) >= I(2 * i + j);
+O.add_constraints({j<2});
+}
+```
+
+with `N = 5`, the indices `i = 1, j = 1` are valid indices.
+However, `i = 2, j = 1` are not valid indices for this operation, nor are `i = -1000, j = 1`.
+
+A set of indices are _valid_ if and only if:
+
+1. All the index variables are integers.
+1. All the index expressions for every tensor are in range. Specifically, if the
+   index variable values are plugged into every index expression, all the
+   resulting indices are non-negative integers less than the appropriate
+   dimension.
+1. All the constraints are satisfied.
+   Constraints always take the form `if ([index expression] < [constant expression])`
```

Comment:
remove the wrapping `if` here, just use "`[index_expression] < [constant expression])`"